addons/mail/models/res_users_settings.py

```diff
@@ -16,6 +26,32 @@ class ResUsersSettings(models.Model):
     voice_active_duration = fields.Integer(string="Duration of voice activity in ms", help="How long the audio broadcast will remain active after passing the volume threshold")
     volume_settings_ids = fields.One2many('res.users.settings.volumes', 'user_setting_id', string="Volumes of other partners")
 
+    @api.constrains('notification_preference')
+    def _check_notification_preference(self):
+        for setting in self:
+            if setting.user_id.share and setting.notification_preference != "email":
+                raise ValidationError(_('Only internal user can receive notifications in Odoo'))
+
+    @api.depends('user_id.share', 'user_id.groups_id')
+    def _compute_notification_preference(self):
```

Comment:
Feel strange to have a business behavior on a technical model. I would move the change on user model, as it is the real model we are working with; probably with an override of create/write as it is more robust than computed fields using many2one.something triggers.